### ISSUE TITLE
Injecting with Dagger 2

### ISSUE THREAD

**Issue by EGecius** [jordanterry/Espresso-Examples#8]
Would you be interested in a sample how in a test Inject objects with Dagger 2 and testing Activity with the injected data? One application of it would be having mock server responses injected into activity.

However, such example would require adding extra dependencies in Gradle, so I wonder if this might confuse potential inexperienced users of this depository as to whether they need Dagger for simple Espresso tests. Probably we could add a clear comment above Dagger dependency clarifying this.

What do you think?

**Comment by jordanterry** [comment]
I would definitely be interested in seeing this. I have been thinking about Dagger in tests recently and it would be great to see what you can do with it.

You are right in saying that inexperienced users might be confused looking at the repo. I think we should have some clear comments as you have mentioned and then add something discussing Dagger 2 to the readme file.

**Comment by jordanterry** [comment]
I was thinking about this earlier and I thought it would be great if we could interact with a free API with an injected service and then when Espresso tests are being run we inject with a mock service that gives us fake data. The perfect use case for Dagger 2 here!

You may have already been thinking about doing this, but I thought I would chirp in!

**Comment by EGecius** [comment]
yeah, makes very good sense. I have nearly finished a similar example but without interacting with a free API. What free API would you suggest?

Another question is how far along do we want to go with this? For communicating with server APIs I would normally use Retrofit but adding it as well may get the project a bit too far-reaching. Ideally, we would keep a project very focused, i.e. do only one thing and do it well.

 What are your thoughts?

**Comment by jordanterry** [comment]
What is your current example? I was thinking maybe [Open Weather Map](http://openweathermap.org/api), it's a free API for getting the weather. I would use Retrofit too for API interactions.

You're right it might be a bit too far reaching, because if we were to add this it would now be about testing in general. Let's keep it to your example so far, we can maybe look at APIs later. Let's do one thing and do it well and then maybe expand upon it in the future.

**Comment by TWiStErRob** [comment]
I don't think Retrofit is needed here. To demonstrate that injection works in an example like this I would just have a class which could call Retrofit, but instead of actually calling anything:
```java
// simulate a long running call, normally this would be a network request with e.g. Retrofit
Thread.sleep(1500);
return Arrays.asList(some, live, data);
```
The above is the "production code". Obviously you want to replace with this in test with
```java
return Arrays.asList(some, mock, data);
```
If the replacement is done without knowing what the original impl is, then it should demonstrate the mocking well. Notice also that I removed the sleep, so it should be visible if the mocking works in addition to the passing test.
